File: spigot/spigot.c
/**
 * Name: Vicky Mohammad
 * Email: mohammav@uoguelph.ca
 */
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

/**
 * Spigot function to find PI.
 * @return malloc string of the result. Should be freed.
 */
char* spigot(int numOfDec){
	char* toBeReturn = calloc(1024, sizeof(char));
	strcpy(toBeReturn, "0");
	// Declare varibles
	int array[numOfDec + 1];
    int i, k, b, d, c = 0;

	// For loop and initialize the array
    for (i = 0; i < numOfDec; i++) {
        array[i] = 2000;
    }// End for

	// For loop
    for (k = numOfDec; k > 0; k -= 14) {
        d = 0;
        i = k;
		// Loop until i equal to 0
        while (1) {
            d += array[i] * 10000;
            b = 2 * i - 1;
			// Calculate and find the remainder
            array[i] = d % b;
            d /= b;
            i--;
            if (i == 0) break;
            d *= i;
        }//End while
        sprintf(toBeReturn, "%s%.4d", toBeReturn, c + d / 10000);
		toBeReturn = realloc(toBeReturn, sizeof(char) * (strlen(toBeReturn) + 8));
        c = d % 10000;
    }// End for loop
	
	return toBeReturn;
}// End spigot function

/**
 * Write to an outout file
 * @filename for the output
 * @string to be outputed
 */
void write(char* fileName, char* string){
	printf("fileName = |%s|\n", fileName);
	FILE* file = fopen(fileName,  "w");
	fprintf(file, "%s", string);
	fclose(file);
}// End file

/**
 * Main function to run the code.
 */
int main(int argc, char** argv){
	// Get the output file name
	char* fileName = calloc(256, sizeof(char));
	printf("Enter the filename to output: \n");
	scanf("%s", fileName);

	// Write the output string to a file
	char* outputString = spigot(1000);
	printf("Writing...\n%s\n", outputString);
	write(fileName, outputString);

	// Free and return
	free(fileName);
	free(outputString);
	return 0;
}// End main function 

File: spigot/spigot.py
# Spigot function
# @param numOfDec
# @return result string
def spigot(numOfDec = 1000):
    # declare variable
    array = []
    toBeReturn = ''
    c = 0
    
    # For loop and initialize the array
    for x in range(0, numOfDec+1):
        array.insert(x, 2000)
    
    # For loop
    k = numOfDec
    while k > 0:
        d = 0
        i = k
        # Loop until i equal to 0
        while True:
            d += array[i] * 10000
            b = 2 * i - 1
            array.insert(i, (d % b))
            d /= b
            i -= 1
            if (i == 0): 
                break
            d *= i
        # Print each digittt
        toBeReturn = toBeReturn + str(c + d / 10000)
        c = d % 10000
        # k value changed
        k = k - 14
        # Return the string output
    return toBeReturn

# Write to an outout file
# @filename for the output
# @string to be outputed
def write(filename, string):
    file = open(filename, "w")
    file.write(string)
    file.close

# Main function to run the program
# for organizing the code.
print("Enter the filename to output: ")
val = raw_input()
result = spigot()
write(val, result)